src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -21,9 +23,37 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceProvider
 
         public async Task Invoke(HttpContext httpContext)
         {
-            using (var container = RequestServicesContainer.EnsureRequestServices(httpContext, _services))
+            // All done if we already have a request services
+            if (httpContext.RequestServices != null)
             {
-                await _next.Invoke(httpContext);
+                return;
+            }
+
+            // Resolve the ScopeFactory from the correct SP
+            var priorApplicationServices = httpContext.ApplicationServices;
+            var serviceProvider = priorApplicationServices ?? _services;
+            var appServiceProvider = serviceProvider.GetRequiredService<IServiceProvider>();
+            if (serviceProvider != appServiceProvider)
+            {
+                appServiceProvider = serviceProvider;
+            }
+            var appServiceScopeFactory = appServiceProvider.GetRequiredService<IServiceScopeFactory>();
+
+            try
+            {
+                // Creates the scope and temporarily swap services
+                using (var scope = appServiceScopeFactory.CreateScope())
+                {
+                    httpContext.ApplicationServices = appServiceProvider;
+                    httpContext.RequestServices = scope.ServiceProvider;
+
+                    await _next.Invoke(httpContext);
+                }
+            }
+            finally
+            {
+                httpContext.RequestServices = null;
+                httpContext.ApplicationServices = priorApplicationServices;
```

Comment:
So now we return RequestServices to null, and ApplicationServices to the old context.ApplicationServices, instead of returning it to the effective serviceProvider which we were doing previously (I don't think this behavior actually matters to anyone except during cleanup (or if someone wraps our IStartupFilter with their own which adds middleware before us)